Add LoRaWAN fragment decoder with lower memory footprint

luqasn opened the pull request:
which uses a recovery matrix which size is based on the fragment loss
tolerance as opposed to the max fragment count in the existing
implementation from semtech's LoRaMac.

This implementation is hopefully going to make forward error correction
usable on lower memory devices.

Some stats from running the FUOTA sample:

### rak4631 (SOC: nRF52840) with min/max frag size 48 bytes:

| Decoder | Redundancy | Resulting RAM usage |
|---------|------------|---------------------|
| Semtech | 13 %       | 259904 (99.15%)     |
| Low Mem | 13 %       | 163072 (62.21%)     |
| Low Mem | 18 %       | 260992 (99.56%)     |

### nucleo_wl55jc (SOC: STM32WL55JC) with min/max frag size 48 bytes:

| Decoder | Redundancy | Resulting RAM usage |
|---------|------------|---------------------|
| Semtech | 18 %       | 63360 (96.68%)      |
| Low Mem | 18 %       | 49728 (75.88%)      |
| Low Mem | 24 %       | 64384 (98.24%)      |

luqasn committed:
tests: subsys: lorawan: run frag decoder test with both implementations

in order to make sure they work.

Signed-off-by: Lucas Romero <luqasn@gmail.com>

martinjaeger commented on subsys/lorawan/services/frag_decoder_lowmem.h:
```diff
@@ -0,0 +1,53 @@
+#define FRAG_DEC_ERR_TOO_MANY_FRAME_LOST (-3)
+#define FRAG_DEC_ERR                     (-4)
+
+typedef struct {
+	/** number of fragments */
+	uint16_t nb_frag;
+	uint8_t frag_size;
+} frag_dec_cfg_t;
+
+typedef enum {
+	/* wait uncoded fragmentations */
+	FRAG_DEC_STA_UNCODED,
+	/* wait coded fragmentations, uncoded frags are processed as coded ones */
+	FRAG_DEC_STA_CODED,
+	FRAG_DEC_STA_DONE,
+} frag_dec_sta_t;
+
+typedef struct {
+	frag_dec_cfg_t cfg;
+	frag_dec_sta_t status;
+
+	uint16_t lost_frame_count;
+	uint16_t filled_lost_frm_count;
+} frag_dec_t;
```
We typically avoid using typedefs for `struct`s and `enum`s in Zephyr (see [here](https://www.kernel.org/doc/html/v4.14/process/coding-style.html#typedefs) for some good reasons). Also I think we could avoid the additional struct for config and just put the two members into the decoder struct:

```suggestion
enum frag_decoder_status {
	/* wait for uncoded fragments */
	FRAG_DEC_STATUS_UNCODED,
	/* wait for coded fragments */
	FRAG_DEC_STATUS_CODED,
	FRAG_DEC_STATUS_DONE,
};

struct frag_decoder {
	uint16_t nb_frag;
	uint8_t frag_size;
	enum frag_decoder_status status;
	uint16_t lost_frame_count;
	uint16_t filled_lost_frm_count;
};
```

martinjaeger commented on subsys/lorawan/services/frag_decoder_lowmem.h:
```diff
@@ -0,0 +1,53 @@
+
+#define FRAG_DEC_ONGOING                 (-1)
+#define FRAG_DEC_ERR_INVALID_FRAME       (-2)
+#define FRAG_DEC_ERR_TOO_MANY_FRAME_LOST (-3)
```
```suggestion
#define FRAG_DEC_ERR_TOO_MANY_FRAMES_LOST (-3)
```

martinjaeger reviewed (changes requested):
Minor comment: The below line should also use the new Kconfig `CONFIG_LORAWAN_FRAG_TRANSPORT_DECODER_SEMTECH`:
https://github.com/zephyrproject-rtos/zephyr/blob/e3866a969071be59fd70e896cfb9970cc9c397f0/modules/loramac-node/CMakeLists.txt#L58

And could you please update the last commit message (or split the commit into two) as you have also added some new tests (which is great) and not just added a new testcase for the new decoder?

martinjaeger commented:
@JiapengLi This PR uses [your original implementation](https://github.com/JiapengLi/LoRaWANFragmentedDataBlockTransportAlgorithm) as a basis. Maybe you want to have a look at this PR and give some review comments? Most important change is that the Zephyr bitarray type is used instead of your custom bitmap implementation.

luqasn committed:
tests: subsys: lorawan: add more test cases

for different numbers of lost fragments.

Signed-off-by: Lucas Romero <luqasn@gmail.com>

luqasn commented:
Thanks @martinjaeger, I have addressed your comments and also fixed some inconsistent naming and comments along the way.
Please have another look.

martinjaeger commented on subsys/lorawan/services/Kconfig:
```diff
@@ -65,6 +65,22 @@ config LORAWAN_FRAG_TRANSPORT
 
 	  The used default port for this service is 201.
 
+choice LORAWAN_FRAG_TRANSPORT_DECODER
+	bool "Fragmented Data Block Transport decoder implementation"
+	default LORAWAN_FRAG_TRANSPORT_DECODER_SEMTECH
```
Looks like you need to add a dependency to fix the CI error:

```suggestion
choice LORAWAN_FRAG_TRANSPORT_DECODER
	bool "Fragmented Data Block Transport decoder implementation"
	depends on LORAWAN_FRAG_TRANSPORT
	default LORAWAN_FRAG_TRANSPORT_DECODER_SEMTECH
```

JordanYates commented on subsys/lorawan/services/frag_decoder_lowmem.c:
```diff
@@ -0,0 +1,411 @@
+SYS_BITARRAY_DEFINE_STATIC(matched_lost_frm_bm1, FRAG_TOLERANCE);
+SYS_BITARRAY_DEFINE_STATIC(matrix_line_bm, FRAG_MAX_NB);
+
+static inline size_t matrix_location_to_index(size_t x, size_t y, size_t m);
```
can all of these forward declarations be avoided if `frag_dec` is moved to the end of the file?

luqasn replied:
Yes, no problem. I did the forward declaration stuff because I wanted the most important implementation first in the file, but that's a matter of taste and I don't mind changing it.

JordanYates commented on subsys/lorawan/services/frag_decoder_lowmem.c:
```diff
@@ -0,0 +1,411 @@
+ *
+ * @returns Pseudorandom output value
+ */
+static int32_t prbs23(int32_t seed);
```
The parameter is not a seed value in the traditional sense, I think `previous` is more accurate.
A PRBS generates the next value from the previous.

JordanYates commented on subsys/lorawan/services/frag_decoder_lowmem.c:
```diff
@@ -0,0 +1,411 @@
+	for (size_t nb_coeff = 0; nb_coeff < (m / 2); nb_coeff++) {
+		r = (1 << 16);
+		while (r >= m) {
+			x = prbs23(x);
```
`x` is undefined on the first run through

luqasn replied:
Bit hard to spot, but it is initialized from `n` here: https://github.com/zephyrproject-rtos/zephyr/pull/73207/files/baee8fa2ba46c85186c38a3bda7ae07e873ab8be#diff-562bcad733f2caab662dfe65f9667d5e0ffe46fc9dea4a1aefdb857cc032fce7R349

JordanYates commented on subsys/lorawan/services/frag_decoder_lowmem.c:
```diff
@@ -0,0 +1,411 @@
+ * Copyright (c) 2022 Jiapeng Li
+ *
+ * Based on: https://github.com/JiapengLi/LoRaWANFragmentedDataBlockTransportAlgorithm
+ *
```
Probably worth including a link to the original algorithm paper linked in the docs: http://www.inference.org.uk/mackay/gallager/papers/ldpc.pdf

JordanYates reviewed (commented):
Working through google translate on the [docs](https://jiapeng.me/lorawan-fragmented-data-block-transport-algorithm/), I'm curious if you have any knowledge about:

> The problem caused by algorithm 2 is that when the number of unencoded data packets is too large, the packets will not be reconstructed.

I'm guessing that is related to the `T` parameter discussed, but I don't see any new parameter introduced in these configs?

JordanYates commented:
If it's not too much work could you split the first commit in two?
1. Moving all the frag_transport code around to get ready for a second decoder
2. Adding the second decoder

It's currently hard to tell whether this PR introduces any changes when using the current decoder

luqasn committed:
lorawan: services: frag transport: refactor status

to no longer be a long-lived status variable because we already map it into
is_active and this reduces the number of states the transport can be in.

It also helps us prepare for being able to plug in more decoders by
removing implementation specific bits from the general transport interface.

Signed-off-by: Lucas Romero <luqasn@gmail.com>

luqasn committed:
lorawan: services: frag transport: prepare for pluggable decoder

by wrapping decoder implementation specific bits in #ifdefs

Signed-off-by: Lucas Romero <luqasn@gmail.com>

luqasn committed:
lorawan: services: frag decoder: add lowmem implementation

that utilizes the fact that the ldpc recovery matrix is triangular and only
stores the half with non-zero values.

This implementation is hopefully going to make forward error correction
usable on lower memory devices.

Signed-off-by: Lucas Romero <luqasn@gmail.com>

luqasn commented:
> If it's not too much work could you split the first commit in two?
> 
> 1. Moving all the frag_transport code around to get ready for a second decoder
> 2. Adding the second decoder
> 
> It's currently hard to tell whether this PR introduces any changes when using the current decoder

Hey again, good remark. It wasn't too much work and it lead to me re-thinking and improving the code slightly by removing a dependency of the transport status on the decoder implementation-specific status codes.

Please have another look.
